### PR TITLE
#163426626 User Reset password

### PR DESCRIPTION
**Pull request by obasajujoshua31**
#### Description of Task to be completed?
Reset user password
 #### How should this be manually tested?
Clone the repo, cd into 
routes 
-  POST /api/v1/users/resetPassword
-PUT /api/v1/users/resetPassword/:token

 #### Any background context you want to provide?
User must have an account already
 #### What are the relevant pivotal tracker stories?
[#163426626](https://www.pivotaltracker.com/story/show/163426626)
 #### Screenshots (if appropriate)
screenshot 2019-02-14 at 5 46 28 pm

 #### Questions:

### CODE REVIEW

File: server/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+const User = database["Users"];
+
+export const sendPasswordReset = async (req, res, next) => {
+  const userEmail = req.body.email;
+  try {
+    const verifiedUser = await User.findOne({
+      where: {
+        email: userEmail
+      }
```
Comment (eokwukwe) [inline server/controllers/users.js]: I thought we check a verified user by checking the `isVerified` field in the database. Using the email alone does not prove that the user is verified.

File: server/controllers/users.js [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      });
+    }
+    const {
+      dataValues: { id, firstName, lastName, email }
+    } = verifiedUser;
+    const userResetToken = generateRandomString(64);
+    const resetUrl = getBaseUrl(req)
+      .concat("/users/resetPassword/")
+      .concat(`${id}/`)
+      .concat(userResetToken);
```
Comment (eokwukwe) [inline server/controllers/users.js]: Can't we use the normal `template string` to achieve the same purpose instead of `concat`?

File: server/controllers/users.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
+import generateRandomString from "crypto-random-string";
+import database from "../models/index";
+import { getBaseUrl, httpResponse, serverError } from "../helpers/index";
+import { sendResetEmail } from "../emails/email";
+import userResponse from "../helpers/userResponse";
+
+const User = database["Users"];
+
+export const sendPasswordReset = async (req, res, next) => {
+  const userEmail = req.body.email;
```
Comment (eokwukwe) [inline server/controllers/users.js]: Prefer object destructuring e.g. `const { email } = req.body`

File: server/controllers/users.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import generateRandomString from "crypto-random-string";
+import database from "../models/index";
+import { getBaseUrl, httpResponse, serverError } from "../helpers/index";
+import { sendResetEmail } from "../emails/email";
+import userResponse from "../helpers/userResponse";
+
+const User = database["Users"];
+
+export const sendPasswordReset = async (req, res, next) => {
```
Comment (eokwukwe) [inline server/controllers/users.js]: Please remove `next`. It is doing nothing here.

File: server/controllers/users.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      .concat(`${id}/`)
+      .concat(userResetToken);
+    await verifiedUser.update({
+      resetToken: userResetToken
+    });
+    sendResetEmail({ firstName, lastName, userEmail: email, resetUrl });
+    return httpResponse(res, {
+      statusCode: 200,
+      message: `Email Sent Successfully to your email, `.concat(
+        "check your Spam in case you did not find it in your inbox"
```
Comment (eokwukwe) [inline server/controllers/users.js]: Why `concat`?

File: server/emails/email.js [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import sgMail from "@sendgrid/mail";
+import { sendVerificationTemplate } from "./templates/verification";
+import { passwordResetTemplate } from "./templates/passwordReset";
+
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+
+export const sendVerificationEmail = async options => {
+  const { firstName, lastName, userEmail, verificationUrl } = options;
+  const msg = {
```
Comment (eokwukwe) [inline server/emails/email.js]: Variable name not descriptive. Please rename.

File: server/emails/email.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    subject: "Verify Email",
+    html: sendVerificationTemplate(firstName, lastName, verificationUrl)
+  };
+  const mail = await sgMail.send(msg);
+  return mail;
+};
+
+export const sendResetEmail = async options => {
+  const { firstName, lastName, userEmail, resetUrl } = options;
+  const msg = {
```
Comment (eokwukwe) [inline server/emails/email.js]: Variable name not descriptive. Please rename.

File: server/helpers/index.js [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+export const getBaseUrl = httpObj => {
+  return `${httpObj.protocol}://${httpObj.get("host")}/api/v1`;
+};
+
+export const generateUrl = options => {
+  const { userId, path, httpObj } = options;
+  return getBaseUrl(httpObj)
+    .concat(path)
+    .concat(`/${generateJWT(userId)}`);
```
Comment (eokwukwe) [inline server/helpers/index.js]: Why `concat`?

**Review by eokwukwe — CHANGES_REQUESTED** [review]
Please check your code. Also your commit message header
`feat(password reset): Users should be able to reset their password` should be
`feat(passwordReset): users should be able to reset their password`.
Please review the wiki for commit messages.

File: server/emails/email.js (hunk 6 above)
Comment (eokwukwe) [reply server/emails/email.js]: `msg` is not a descriptive variable name. Please rename the variable.

File: server/emails/email.js (hunk 7 above)
Comment (eokwukwe) [reply server/emails/email.js]: `msg` is not a descriptive variable name. Please rename the variable.

File: server/emails/templates/passwordReset.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      action: {
+        instructions: "Click the button below to reset your password:",
+        button: {
+          color: "#DC4D2F",
+          text: "Reset your password",
+          link: resetUrl
+        }
+      },
+      outro: `If you did not request a password reset`.concat(
+        "further action is required on your part."
```
Comment (eokwukwe) [inline server/emails/templates/passwordReset.js]: Why are you still using `concat`?

File: server/controllers/user.js [hunk 10]
```diff
@@ -40,11 +41,61 @@ export const userLogin = async (req, res) => {
...
+};
+
+export const passwordResetRequest = async (req, res) => {
+  const { email } = req.body;
+  try {
+    const verifiedUser = await User.findOne({
+      where: {
+        email
+      }
+    });
```
Comment (eokwukwe) [inline server/controllers/user.js:56]: Does email alone indicate that a user is `verified`. You need to include a check that the user is a verified user.
```
const verifiedUser = await User.findOne({
  where: {
    email,
    isVerified: true
  }
})
```

File: server/controllers/user.js [hunk 11]
```diff
@@ -40,11 +41,61 @@ export const userLogin = async (req, res) => {
...
+    if (verifiedUser) {
+      const url = `${getBaseUrl(req)}/users/resetPassword`;
+      await verifiedUser.sendPasswordResetEmail(url);
+      return httpResponse(res, {
+        statusCode: 200,
+        message: `Email Sent Successfully to your email, `.concat(
+          "check your Spam in case you did not find it in your inbox"
```
Comment (eokwukwe) [inline server/controllers/user.js:64]: No `concat` please

**Review by eokwukwe — CHANGES_REQUESTED** [review]
You have not also edited your commit message. I pointed it out before: 
Please check your commit message header
`feat(password reset): Users should be able to reset their password` should be
`feat(passwordReset): users should be able to reset their password`.
Please review the wiki for commit messages.

**Review by eokwukwe — APPROVED** [review]

**Review by ghost — APPROVED** [review]

File: package.json [hunk 12]
```diff
@@ -49,6 +49,7 @@
     "lodash.isempty": "^4.4.0",
     "mailgen": "^2.0.7",
     "methods": "^1.1.2",
+    "moment": "^2.24.0",
```
Comment (benfluleck) [inline package.json]: Why did you install this ?

File: server/configs/config.js [hunk 13]
```diff
@@ -14,8 +15,10 @@ const config = {
   },
   production: {
     url: process.env.PROD_DATABASE_URL,
-    dialect: process.env.PROD_DIALECT
+    dialect: process.env.PROD_DIALECT,
+    logging: false
   }
 };
 
+export const TTL = process.env.NODE_ENV === "test" ? 5 : process.env.TTL;
```
Comment (benfluleck) [inline server/configs/config.js]: What is this, please give a more descriptive name

File: server/controllers/user.js [hunk 14]
```diff
@@ -2,10 +2,12 @@ import isEmpty from "lodash.isempty";
...
 import { validateData } from "../validations/validateData";
 import { signInSchema } from "../validationSchemas/user";
-import { generateJWT } from "../helpers/jwt";
 import { comparePassword } from "../helpers/password";
-import { httpResponse } from "../helpers/http";
+import { httpResponse, serverError } from "../helpers/http";
+import { getBaseUrl } from "../helpers/index";
+import userResponse from "../helpers/userResponse";
 
+const User = database["Users"];
```
Comment (benfluleck) [inline server/controllers/user.js]: import Users from models, we should be using this syntax again

File: server/emails/email.js [hunk 15]
```diff
@@ -0,0 +1,29 @@
+import sgMail from "@sendgrid/mail";
```
Comment (benfluleck) [inline server/emails/email.js:1]: This is a repetition on @teepha  logic, Please refactor an use methods so we keep our code DRY

File: server/emails/templates/passwordReset.js [hunk 16]
```diff
@@ -0,0 +1,31 @@
+import Mailgen from "mailgen";
+
+const mailGenerator = new Mailgen({
+  theme: "default",
+  product: {
```
Comment (benfluleck) [inline server/emails/templates/passwordReset.js]: This is a repetition of @teepha  logic

File: server/helpers/http.js [hunk 17]
```diff
@@ -16,3 +17,22 @@ export const httpResponse = (res, options) => {
...
+ * @param {object} error error object
+ * @returns {object} HTTP response
+ * @method serverError
+ */
+export const serverError = (res, error) => {
+  console.log(error.message);
+  return res.status(error.status || 500).json({
+    success: false,
+    message:
+      error && cleanString(process.env.NODE_ENV) === "production"
```
Comment (benfluleck) [inline server/helpers/http.js]: This line is not very readable. I have told @mrkingz  to remove this function and you were right beside me when I was doing it.

better 
`const isProudction = process.env.NODE_ENV === "production"`

now we can do 
```
message:
      error && isProduction
```

File: server/helpers/index.js [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+/**
+ * @description Removes all white space character, if any
+ *
+ * @method cleanString
+ * @param {Strng} string The string to remove white space(s)
+ * @param {Boolean} removeAll: if false, multiple white space will reduce to single;
+ * @returns {Object} An object with trimed attributes
+ * @method cleanString
+ */
+export const cleanString = (string = "", removeAll) => {
```
Comment (benfluleck) [inline server/helpers/index.js]: This function is very poorly written with too many difficult to understand variable names and unnecessary checks.

File: server/helpers/passportCallback.js [hunk 19]
```diff
@@ -1,8 +1,10 @@
 import userResponse from "./userResponse";
 import database from "../models/index";
+import socialAuthConfig from "../configs/passport";
 
 const User = database["Users"];
```
Comment (benfluleck) [inline server/helpers/passportCallback.js]: import User from 'models'

File: server/helpers/passportCallback.js [hunk 20]
```diff
@@ -1,8 +1,10 @@
...
-const performCallback = (token, tokenSecret, profile, done) => {
+
+export const performCallback = (token, tokenSecret, profile, done) => {
```
Comment (benfluleck) [inline server/helpers/passportCallback.js]: Rename this method, this is not very descriptive enough

File: server/emails/templates/passwordReset.js [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+      action: {
+        instructions: "Click the button below to reset your password:",
+        button: {
+          color: "#DC4D2F",
+          text: "Reset your password",
+          link: resetUrl
+        }
+      },
+      outro: `If you did not request a password reset
+        further action is required on your part.`
```
Comment (ghost) [inline server/emails/templates/passwordReset.js]: This line is displaying this in the email sample you attached: `If you did not request a password resetfurther action is required on your part.`. Please add ` No` before the `further action`

File: server/helpers/passportCallback.js [hunk 22]
```diff
@@ -29,4 +32,7 @@ const performCallback = (token, tokenSecret, profile, done) => {
     }
   });
 };
-export default performCallback;
+
+export const getAuthCredentials = platform => {
```
Comment (benfluleck) [inline server/helpers/passportCallback.js]: getOauthCredentials I think this name is better

File: server/models/users.js [hunk 23]
```diff
@@ -71,7 +87,36 @@ const getUserModel = (sequelize, Sequelize) => {
...
       onUpdate: "CASCADE"
     });
+    User.hasMany(db["Articles"], {
+      foreignKey: "authorId",
+      target: "id",
+      onDelete: "CASCADE"
+    });
+  };
+
+  User.prototype.generateResetToken = function() {
```
Comment (benfluleck) [inline server/models/users.js:98]: Why on the prototype and not on the function?. Also I was able to use arrow functions in my helloBooks project. Go an check it out

File: server/tests/index.spec.js [hunk 24]
```diff
@@ -2,7 +2,8 @@ import "./model/articles.specs";
 import "./model/usersModel.specs";
 import "./model/favourites.specs";
 import "./miscellaneous/connection.spec";
-import "./miscellaneous/emails/verification/email.specs";
 import "./oauth/social.auth.spec";
-import "./controller/userController";
+import "./miscellaneous/emails/verification";
+import "./miscellaneous/emails/resetPassword";
 import "./validations/user.spec";
+import "./routes/users.spec";
```
Comment (benfluleck) [inline server/tests/index.spec.js]: Why did you change this?  change it back

File: server/tests/routes/users.spec.js [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+      .send({ email: "none@example.com" });
+    expect(response.statusCode).to.equal(404);
+  });
+
+  it("should successfully send a link to a registered User with a status code of 200", async () => {
+    await User.create(validUser);
+    const response = await chai
+      .request(app)
+      .post(`${baseUrl}`)
+      .send({ email: "obasajujoshua31@gmail.com" });
```
Comment (benfluleck) [inline server/tests/routes/users.spec.js]: Please split this test it is too big. Unit test should be small and compact

**Review by benfluleck — CHANGES_REQUESTED** [review]

File: server/helpers/index.js (hunk 18 above)
Comment (ghost) [inline server/helpers/index.js]: Benny left feedback about this `cleanString` function on Kingsley's PR. Please check it and implement it here as well

File: package.json (hunk 12 above)
Comment (obasajujoshua31) [reply package.json]: I am working with time stamp

Comment (benfluleck) [reply package.json]: There are much smaller libraries available. Did you care to check the size of the library on your bundle

File: server/controllers/user.js (hunk 14 above)
Comment (obasajujoshua31) [reply server/controllers/user.js]: I will have to refractor the user model to be able to do that. I tried it first, it didn't work, it will break every body work that have used the user model

File: server/configs/config.js [hunk 26]
```diff
@@ -14,8 +15,11 @@ const config = {
   },
   production: {
     url: process.env.PROD_DATABASE_URL,
-    dialect: process.env.PROD_DIALECT
+    dialect: process.env.PROD_DIALECT,
+    logging: false
   }
 };
 
+export const expiryTime = process.env.NODE_ENV === "test" ? 7 : process.env.TTL;
```
Comment (benfluleck) [inline server/configs/config.js]: expiration for What?
Don't hard code variables store them in constants.
What is TTL?

File: server/emails/email.js (hunk 15 above)
Comment (benfluleck) [reply server/emails/email.js:1]: Please refactor this logic it is too tightly coupled and does not leave room for expansion

File: .env.example [hunk 27]
```diff
@@ -26,4 +26,7 @@ JWT_SECRET=
 
 PORT=
 EMAIL=me@example.com
-SENDGRID_API_KEY=kdkkdkdkkdkd
\ No newline at end of file
+SENDGRID_API_KEY=kdkkdkdkkdkd
+
+EXPIRY_TIME
+EXPIRY_TIME_TEST
```
Comment (benfluleck) [inline .env.example]: New line needed here

File: server/emails/email.js [hunk 28]
```diff
@@ -0,0 +1,27 @@
...
+import emailTemplate from "./emailTemplate";
+
+sgMail.setApiKey(process.env.SENDGRID_API_KEY);
+
+// /**
```
Comment (benfluleck) [inline server/emails/email.js]: These comments are incorrectly formatted

File: server/emails/emailTemplate.js [hunk 29]
```diff
@@ -0,0 +1,43 @@
...
+
+const mailGenerator = new Mailgen({
+  theme: "default",
+  product: {
+    name: "Authors Haven",
+    link: "#"
+  }
+});
+
+// Prepare email contents
```
Comment (benfluleck) [inline server/emails/emailTemplate.js]: Remove this comment  or put it in your jsdoc

File: server/emails/emailTemplate.js [hunk 30]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * @param  {String} firstName
+ * @param  {String} lastName
+ * @param  {String} url
+ * @methods passwordResetTemplate
+ */
+
+const emailTemplate = (
```
Comment (benfluleck) [inline server/emails/emailTemplate.js]: If this is a method name it like a method e.g.` generateEmailTemplate` or `getEmailTemplate`

File: server/helpers/crypto.js [hunk 31]
```diff
@@ -0,0 +1,34 @@
...
+  );
+  let decodedTime = timeDecoded.update(token, "hex", "utf8");
+  decodedTime += timeDecoded.final("utf8");
+
+  return decodedTime;
+};
+
+export const getTimeDifference = time => {
+  const timeIssued = new Date(time);
+  const now = new Date();
```
Comment (benfluleck) [inline server/helpers/crypto.js]: Rename this variable please.

File: server/models/users.js [hunk 32]
```diff
@@ -71,8 +88,42 @@ const getUserModel = (sequelize, Sequelize) => {
...
+    this.resetToken = encryptToken();
+    this.save();
+    this.reload();
+    return this.resetToken;
   };
+
+  User.prototype.sendPasswordResetEmail = function(url) {
```
Comment (benfluleck) [inline server/models/users.js:105]: I don't believe you should be modifying the prototypes to be implementing the hooks. This seems like a hack. Please re-implement this

File: server/tests/controller/userController/resetPassword.spec.js [hunk 33]
```diff
@@ -0,0 +1,75 @@
+import chai, { expect } from "chai";
+import chaiHttp from "chai-http";
+import app from "../../../index";
+import models from "../../../models";
+import { validUser } from "../../fixtures/model/userData";
+
+const { Users } = models;
+
+const timeout = (func, delay) => {
```
Comment (benfluleck) [inline server/tests/controller/userController/resetPassword.spec.js]: Please rename it like a function. e.g. `getTimeout`

**Review by benfluleck — CHANGES_REQUESTED** [review]
